### PR TITLE
[BISERVER-14912] - Creating schedules with “Run Once“ recurrence in Deutsch language does not work in PUC.

### PR DESCRIPTION
**Pull request by andreramos89**
@smmribeiro 
@bcostahitachivantara 
@renato-s

### CODE REVIEW

File: widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java [hunk 1]
```diff
@@ -48,4 +48,12 @@ public boolean isValid() {
 
   public void clear() {
   }
+
+  private String normalizeDateEn (String dateString){
+    if(dateString != null && !dateString.isEmpty()){
+      return dateString.replaceAll( "vorm." , "am").replaceAll( "nachm." , "pm") ;
```
[the remark this answers is not in the document]
Comment (smmribeiro) [reply widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java:54]: Also, I'm very curious: is it only in german that one gets words different from AM/PM?

**Review by smmribeiro — CHANGES_REQUESTED** [review]

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 6m 37s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl widgets
```
---
:ok_hand: All tests passed!
 
Tests run: **330**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/gwt-widgets/job/PR-879/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java [hunk 2]
```diff
@@ -48,4 +48,12 @@ public boolean isValid() {
...
   public void clear() {
   }
+
+  private String normalizeDateEn (String dateString){
+    if(dateString != null && !dateString.isEmpty()){
+      return dateString.replace( "vorm." , "am").replace( "nachm." , "pm") ;
+    }else {
+      return dateString;
+    }
+  }
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java:58]: Formatting is off.

**Review by dcleao — CHANGES_REQUESTED** [review]

File: widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java (hunk 2 above)
Comment (dcleao) [reply widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java:58]: Also, this implementation looks to be, essentially, a hack. Could a locale-aware date function be used instead?

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 9m 29s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl widgets
```
---
:ok_hand: All tests passed!
 
Tests run: **398**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/gwt-widgets/job/PR-879/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java [hunk 3]
```diff
@@ -35,17 +36,41 @@ public boolean isValid() {
...
-      final String dateTime = date + " " + editor.getStartTime(); //$NON-NLS-1$
+      String time = editor.getStartTime();  //format of time is "hh:mm:ss a"
+
+      if (editor.getStartDate().before( new Date() )){
+       isValid=false;
+      } else {
+
+        if(time.endsWith( TimeUtil.TimeOfDay.AM.toString() ) || time.endsWith( (TimeUtil.TimeOfDay.PM.toString() ) )){
+          time = normalizeTime( time );
+        }
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java:50]: Test for presence of AM/PM is already done inside `normalizeTime`. There's no need to repeat, afaict.

File: widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java [hunk 4]
```diff
@@ -35,17 +36,41 @@ public boolean isValid() {
     if ( null == editor.getStartDate() ) {
       isValid = false;
     } else {
+
       final DateTimeFormat format = DateTimeFormat.getFormat( "MM-dd-yyyy" ); //$NON-NLS-1$
       final String date = format.format( editor.getStartDate() );
-      final String dateTime = date + " " + editor.getStartTime(); //$NON-NLS-1$
+      String time = editor.getStartTime();  //format of time is "hh:mm:ss a"
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java:42]: Variable can be moved to inside the below `else` case.

File: widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java [hunk 5]
```diff
@@ -35,17 +36,41 @@ public boolean isValid() {
...
+
+  private String normalizeTime( String time ) {
+    if ( time != null && !time.isEmpty() &&
+      ( time.endsWith( TimeUtil.TimeOfDay.AM.toString() ) || time.endsWith( TimeUtil.TimeOfDay.PM.toString() ) ) ) {
+      if ( time.endsWith( TimeUtil.TimeOfDay.AM.toString() ))
+        time = time.replace( TimeUtil.TimeOfDay.AM.toString(), "am");
+      else
+        time = time.replace( TimeUtil.TimeOfDay.PM.toString(), "pm");
+    }
+    return time;
```
Comment (dcleao) [inline widgets/src/main/java/org/pentaho/mantle/client/dialogs/scheduling/validators/RunOnceEditorValidator.java:74]: IMO, would be simpler if the AM/PM tests would only be done by the inner `if`/`else`, while transforming the `else` into an `else if` for PM.

**Review by dcleao — CHANGES_REQUESTED** [review]
Also, formatting is off.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 7m 22s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl widgets
```
---
:ok_hand: All tests passed!
 
Tests run: **398**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/gwt-widgets/job/PR-879/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 7m 6s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl widgets
```
---
:ok_hand: All tests passed!
 
Tests run: **398**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/gwt-widgets/job/PR-879/6/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Review by dcleao — APPROVED** [review]

**Review by smmribeiro — APPROVED** [review]

**Pull request merged**